### PR TITLE
LPS-127344 Add A11y Integration to DXP

### PR DESCRIPTION
**Pull request by matuzalemsteles**
This is just a draft to discuss a few things, still a few things to add and I also made some improvements to the checker to avoid some unnecessary analysis and I refactor the mutation API to simplify its use.

Some points for discussion:

- I'm not a backend expert here but I would like to know if this criterion would be possible to achieve:
  - The tool is not bundled in builds meant to be shipped to customers
  - If the tool is disabled should we still show the settings in the system settings? it is possible?
- Since the tool is enabled should we ignore some pages from being analyzed, for example, system settings?
- By default should we ignore elements like Product Menu, Global Menu, or parts of the infra?

cc @dsanz 

> This still depends on PR #1076

### TODO

- [x] ~~Add mutation deny list criteria to configuration~~
- [x] Add axe-core config to configuration
- [x] Ignore the content area for the editors (any type of editor)

### CODE REVIEW

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	id = "com.liferay.frontend.js.a11y.web.internal.configuration.A11yConfiguration",
+	localization = "content/Language", name = "a11y-configuration-name"
+)
+public @interface A11yConfiguration {
+
+	@Meta.AD(
+		deflt = "false", description = "enable-description",
+		name = "enable-name", required = false
+	)
+	public boolean enable();
```
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:36]: For now I'm adding this flag to disable a11y by default even with nodejs.node.env in development.

File: portal-impl/src/com/liferay/portal/util/PropsValues.java [hunk 2]
```diff
@@ -1895,6 +1895,9 @@
 	public static final String NETVIBES_SERVLET_MAPPING = PropsUtil.get(
 		PropsKeys.NETVIBES_SERVLET_MAPPING);
 
+	public static final String NODEJS_NODE_ENV = GetterUtil.getString(
+		PropsUtil.get(PropsKeys.NODEJS_NODE_ENV));
+
```
Comment (matuzalemsteles) [inline portal-impl/src/com/liferay/portal/util/PropsValues.java:1900]: Apparently, it wasn't exposed but I'm not sure if this was intentional, let me know if there is any problem.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/resources/META-INF/resources/index.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	return container;
+};
+
+const defaultProps = {
+	mutations: {
+		div: {'data-restore-title': ['*'], draggable: ['*']},
+		input: {value: ['*']},
+	},
```
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-a11y-web/src/main/resources/META-INF/resources/index.tsx:39]: I haven't made the annotations to add the mutation configuration, should we move this default configuration there to be visible or do we want things like that just at the code level?

**Comment by nhpatt** [comment]
> The tool is not bundled in builds meant to be shipped to customers

You could stop the module depending on the value of NODE_ENV, yes. The module would be there but stopped/disabled.

> If the tool is disabled should we still show the settings in the system settings? it is possible?

Mmm... you could disable just `A11yTopHeadJSPDynamicInclude` and leave the settings alive and use them to enable/disable `A11yTopHeadJSPDynamicInclude` conditionally.

But I do not know what @dsanz prefers. There is also a common pattern for [disabling modules/components](https://help.liferay.com/hc/en-us/articles/360017895432-Blacklisting-OSGi-Modules-and-Components) that can be switched through the UI or deploying a properties file... so I don't know if it makes more sense to rely on that.

**Comment by matuzalemsteles** [comment]
> You could stop the module depending on the value of NODE_ENV, yes. The module would be there but stopped/disabled.

hey @nhpatt thanks, Do you have any examples for this? 🙂

> Mmm... you could disable just A11yTopHeadJSPDynamicInclude and leave the settings alive and use them to enable/disable A11yTopHeadJSPDynamicInclude conditionally.

Yes, I'm just going to disable `A11yTopHeadJSPDynamicInclude` conditionally based on `nodejs.node.env` but the settings are still visible, I don't know if that's a problem.

File: portal-impl/src/com/liferay/portal/util/PropsValues.java (hunk 2 above)
Comment (dsanz) [reply portal-impl/src/com/liferay/portal/util/PropsValues.java:1900]: This is a build-time property which is in default [build.properties](https://github.com/liferay/liferay-portal/blob/master/build.properties#L509-L511) file. 

The original requirement was set at build time, meaning the tool sources would only package a bundle with the checker if one builds the portal in development mode.

This would prevent our tool from being released.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java (hunk 1 above)
Comment (dsanz) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:36]: Yeah, this makes sense to me as we may want to disable the tool under certain circumstances. I have to say that Chema has a stronger opinion as he'd prefer the tool to be always there.
A reasonable middle ground to me is to have this configuration hidden from the UI, like [this](https://github.com/liferay/liferay-portal/blob/51446053a22ecb2a1d8f6a971d36b2c4c68b6d23/modules/apps/asset/asset-auto-tagger-service/src/main/java/com/liferay/asset/auto/tagger/internal/configuration/AssetAutoTaggerGroupConfiguration.java#L25)

In case we go this way, we need another configuration which should be visible in the system settings to tweak the tool parameters

**Comment by dsanz** [comment]
> > Mmm... you could disable just A11yTopHeadJSPDynamicInclude and leave the settings alive and use them to enable/disable A11yTopHeadJSPDynamicInclude conditionally.
> 
> Yes, I'm just going to disable `A11yTopHeadJSPDynamicInclude` conditionally based on `nodejs.node.env` but the settings are still visible, I don't know if that's a problem.

We may do this but it feels somehow redundant. In case `nodejs.node.env` differs from `development` osgi bundle should just not be there (tomcat bundle you're running does not have it) so there should not be a need to disable it.

**Comment by nhpatt** [comment]
> We may do this but it feels somehow redundant. In case `nodejs.node.env` differs from `development` osgi bundle should just not be there (tomcat bundle you're running does not have it) so there should not be a need to disable it.

Ok, I did not know the requirements. If you are just bundling it in development, the settings are not needed right? or we want to also disable it conditionally in development?

File: portal-impl/src/com/liferay/portal/util/PropsValues.java (hunk 2 above)
Comment (matuzalemsteles) [reply portal-impl/src/com/liferay/portal/util/PropsValues.java:1900]: Hmm Okay, I think I get it, in this case, I'm enabling in runtime instead of build time.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java (hunk 1 above)
Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:36]: Hmm I think I also prefer it not having the option, I just added it here as an attempt to send it before I integrate the Panel with the main component here just as a temp until we have everything in master and more stable.

**Comment by nhpatt** [comment]
This depends on what dani prefers and if it makes sense to expose a `enabled` flag. And if it makes sense to expose the settings or not (with the build UI).

But just for reference, if you want to disable a component you can do it like [this](https://github.com/liferay/liferay-portal/blob/b3173da81b62430f7150b695d5e944b224279580/modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/configuration/persistence/listener/VulcanConfigurationModelListener.java#L70).

**Comment by matuzalemsteles** [comment]
Thanks @dsanz and @nhpatt, I think I just got confused as to how we are going to package this, if I understand the idea is to only package this at build time instead of runtime, so in theory, in production, the settings would not be visible either.

**Comment by dsanz** [comment]
> Oh, I read the other comments, so you maybe want both things (I think it's your decision @dsanz).

Yeah, I think that the best option is to have:
* Build-time flag via node env variable
* Hidden runtime flag which defaults to `enabled`

So, in dev scenarios, tool will be there and requires an non-trivial explicit action (placing a crafted file with a weird name in osgi/configs) to disable.

**Comment by matuzalemsteles** [comment]
Thanks @dsanz, makes sense to me!

**Comment by matuzalemsteles** [comment]
I sent a few more changes to add more configs to the settings and add the flag that lets you disable the tool, I still need to add the axe-core settings and a way to configure the mutations in the settings.

I'm still analyzing how best to deal with mutations, I'm seeing a lot of problems with having long analyzes (800ms to 1000ms), apparently, when we pass the `denylist` along with the target to the axe-core context, it analyzes all the content, I'm investigating to understand better, at the moment I'm not passing the denylist since we're using it to ignore the mutation too, this has served well so far.

As we're target `body`, this has helped identify which mutations we want to ignore and has helped define a better API for that, I'm still working on it to tweak the settings and APIs so that using the tool doesn't have problems with performance.

**Comment by matuzalemsteles** [comment]
Hey @dsanz, I'm working on adding axe-options and mutations to the configuration but I think we have a problem selecting the layout, for example, axe-options has some props, it doesn't make sense that we are adding to the configuration because some properties affect the component's behavior that might break, so I'm adding some configurable properties, and for mutations, I think we need to add a new configuration like a factory, to materialize the mutations API in a UI. So I'm thinking we can add a specific category for the A11y component to organize the settings. What do you think?

**Comment by dsanz** [comment]
I sent a few more changes to add more configs to the settings and add the flag that lets you disable the tool, I still need to add the axe-core settings and a way to configure the mutations in the settings.

> I'm still analyzing how best to deal with mutations, I'm seeing a lot of problems with having long analyzes (800ms to 1000ms), apparently, when we pass the denylist along with the target to the axe-core context, it analyzes all the content,

So, to make sure I understand, is axe-core ignoring the include-exclude object?

> I'm investigating to understand better, at the moment I'm not passing the denylist since we're using it to ignore the mutation too, this has served well so far.

This makes sense to me, still, I think both approaches (ignoring mutations if target is in denylist + add denylist to `config.exclude`) complement each other, so it's great that you're investigating this. In theory, it's possible for a mutation target outside the denylist to have added nodes within the denylist.

> As we're target body, this has helped identify which mutations we want to ignore and has helped define a better API for that, I'm still working on it to tweak the settings and APIs so that using the tool doesn't have problems with performance.

Part of the idea of separating the tool from its integration into DXP is to reach optimum performance while functioning in DXP. So anything we can think of along this line makes sense to me.

> Hey @dsanz, I'm working on adding axe-options and mutations to the configuration but I think we have a problem selecting the layout, for example, axe-options has some props, it doesn't make sense that we are adding to the configuration because some properties affect the component's behavior that might break, so I'm adding some configurable properties, 

I like the approach of providing some "high level" flags which are translated into specific denylist elements. We should have control on what's added to that list, meaning that the config options we expose would basically add stuff to the list, but the user providing the conf should not be able to empty it by just providing []. Reason is that for sure, we want to _ignore ourselves_, meaning that any markup produced by the tool will be in the denylist and there should be no way to circumvent this.

Also, for the rest of axe-core options, some of them may be a bit internal. I don't mind exposing all of them in the first iteration if that helps us to fine tune the behavior (example: don't look into iframes till we have a properly integrated panel, enable perf logs, or just limit the `resultTypes` to `violations`). These may be hidden in the future, or moved to the "hidden" config file.

> and for mutations, I think we need to add a new configuration like a factory, to materialize the mutations API in a UI. So I'm thinking we can add a specific category for the A11y component to organize the settings. What do you think?

Main value provided by the tool is the ability to react to page changes and trigger new analysis, so that developers can see how well the a11y rules are obeyed, w/o having to open a dev tools extension and hit "analyze" all the time. For this, tool must observe page mutations and act accordingly. Some mutations are of interest and some are not, so tool should only analyze a subset of all the changes happening in a page. When integrated in DXP, we're not interested in analyzing what's inside the editors and also what the tool itself produces (overlays, panels, etc...). 

The above can be met if we ignore some mutations. That's exactly what you did in [this commit](https://github.com/liferay-frontend/liferay-portal/pull/1146/commits/33962f0e82608390209ee9ad4b014cc2cdcf3132). Ignoring mutations inside denied elements is a way to meet the requirement above, and should be a good improvement.

I'm wondering what would be the net benefit of having additional mutation criteria to the tool. In other words, which particular additions/deletions do we want to ignore, which are not already in the denylist? and how to properly express them w/o boosting the complexity of the tool config and logic?

I can see some cases. To begin, we should ignore any mutation of type `characterData`, even if the parent element receiving the character data is not denied. It's just that we don't want to re-analyze it on every character data chagne.

Accordingly, the addition/removal of any attributes not related to styling may be ignored, even if the node itself is not in the deny list. We may therefore find a list of attributes which mutations won't trigger the analysis, no matter in which node are they added/removed. I tend to believe that expressing the list of attributes fits better in the form of allow list, however I may be wrong here.

Finally, I'm having a hard time finding cases of mutations in child nodes that do we want to ignore besides the ones in denylist. I would adventure to say that we may want to ignore nodes if they are repetitions of one we already analyzed, such as the nodes in a tree. However, expressing this via mutation crieria is far from trivial as we need to analyze one at least, ignoring the rest. And we have to make sure that all of them represent the same thing and so can be safely ignored, which is an app-level knowledge, too specific even for a DXP integration.

So, to sum up, havinf some mutation criteria makes sense to me. However, if that makes things easier, we don't necessarily need to expose these via UI in the first iteration. We may want to give ourselves some time to explore this need and just start by ignoring some default, well known mutations.

Does this make sense to you?

**Comment by matuzalemsteles** [comment]
Hey @dsanz, makes total sense to me, Thanks for taking the time and answering everything.

> So, to make sure I understand, is axe-core ignoring the include-exclude object?

Yeah, that's the impression I'm getting, apparently, it ignores what's in the `exclude` of the parse but doesn't respect the `include`, I'll check again if there's any edge case, actually, I pass the node reference, it could be that has some implication with that, still investigating in the axe-core codebase.

> This makes sense to me, still, I think both approaches (ignoring mutations if target is in denylist + add denylist to config.exclude) complement each other, so it's great that you're investigating this. In theory, it's possible for a mutation target outside the denylist to have added nodes within the denylist.

Exactly, I took some cases.

> Also, for the rest of axe-core options, some of them may be a bit internal. I don't mind exposing all of them in the first iteration if that helps us to fine tune the behavior (example: don't look into iframes till we have a properly integrated panel, enable perf logs, or just limit the resultTypes to violations). These may be hidden in the future, or moved to the "hidden" config file.

Perfect, it makes sense, I'll go with it.

> I'm wondering what would be the net benefit of having additional mutation criteria to the tool. In other words, which particular additions/deletions do we want to ignore, which are not already in the denylist? and how to properly express them w/o boosting the complexity of the tool config and logic? [...]

Yeah, I get a bit of that feeling as I'm using and configuring, but I've got some use cases for the mutation criteria for adding or removing attributes, for example, we want to ignore any changes to the `id`, I think that's it of the YUI because every time there is a change in the `id`, I remember that the YUI kept creating a unique id for each node. I'm also ignoring any `value` changes of an `input`, there are also many nodes that keep changing the `draggable` property and we don't need to cause an analysis based on that, nor do we want to trigger an analysis based on `data-restore-title`, these were some of what I found.

For mutations coming from adding or removing an element, I also don't see many use cases, what I'm doing now is, when adding a new element, scheduling a new parse for that element, on removal, scheduling a new analysis for the target.

> Accordingly, the addition/removal of any attributes not related to styling may be ignored, even if the node itself is not in the deny list. We may therefore find a list of attributes which mutations won't trigger the analysis, no matter in which node are they added/removed. I tend to believe that expressing the list of attributes fits better in the form of allow list, however I may be wrong here.

Yeah, I think we can have an allow list for that but it would be difficult to list all attributes, for example, we also want to analyze mutations coming from `aria-label`, `role`... it would be a bit difficult to list everything, for now, to go with a deny list might suffice but we can experiment and we can just leave it set at the code level for now.

> Finally, I'm having a hard time finding cases of mutations in child nodes that do we want to ignore besides the ones in denylist. I would adventure to say that we may want to ignore nodes if they are repetitions of one we already analyzed, such as the nodes in a tree. However, expressing this via mutation crieria is far from trivial as we need to analyze one at least, ignoring the rest. And we have to make sure that all of them represent the same thing and so can be safely ignored, which is an app-level knowledge, too specific even for a DXP integration.

Yeah, I thought about creating some heuristic algorithm for that, for example, just schedule an analysis for this element if there isn't an equal element or a higher level element, because we may have higher elements already scheduled but we would have to check if the element that we are scheduling inside within some element in the queue, but this can be a bit of expensive, we can explore this at some point to reduce the number of scheduled elements.

> So, to sum up, havinf some mutation criteria makes sense to me. However, if that makes things easier, we don't necessarily need to expose these via UI in the first iteration. We may want to give ourselves some time to explore this need and just start by ignoring some default, well known mutations.

It makes sense, I think it will be better so we can experiment with what approaches we want to take.

**Comment by dsanz** [comment]
> Hey @dsanz, makes total sense to me, Thanks for taking the time and answering everything.
> 
> Yeah, that's the impression I'm getting, apparently, it ignores what's in the `exclude` of the parse but doesn't respect the `include`, I'll check again if there's any edge case, actually, I pass the node reference, it could be that has some implication with that, still investigating in the axe-core codebase.

> Yeah, I get a bit of that feeling as I'm using and configuring, but I've got some use cases for the mutation criteria for adding or removing attributes, for example, we want to ignore any changes to the `id`, I think that's it of the YUI because every time there is a change in the `id`, I remember that the YUI kept creating a unique id for each node. I'm also ignoring any `value` changes of an `input`, there are also many nodes that keep changing the `draggable` property and we don't need to cause an analysis based on that, nor do we want to trigger an analysis based on `data-restore-title`, these were some of what I found.

Awesome!
 
> For mutations coming from adding or removing an element, I also don't see many use cases, what I'm doing now is, when adding a new element, scheduling a new parse for that element, on removal, scheduling a new analysis for the target.

This makes sense, I think removing nodes may cause some a11y issue, however, no specific example or case comes to my mind at this moment.
 
> Yeah, I thought about creating some heuristic algorithm for that, for example, just schedule an analysis for this element if there isn't an equal element or a higher level element, because we may have higher elements already scheduled but we would have to check if the element that we are scheduling inside within some element in the queue, but this can be a bit of expensive, we can explore this at some point to reduce the number of scheduled elements.

Sure, some ideas about this are in the task if I recall. Maybe we could add some id to the queue elements at least to ease direct comparison to avoid repetition. Avoiding adding a child of an enqueued parent may be more involved. In the end, we'll have to see if it's worth the effort to add some id to the queue or using `compareDocumentPosition()`, `contains()` or similar.

There's also the option of removing from the queue any deleted node reported by a MutationRecord (including children). Again, I'm not sure if this would be better than just dequeue and check if node is in the doc before triggering the analysis...

**Comment by matuzalemsteles** [comment]
> Sure, some ideas about this are in the task if I recall. Maybe we could add some id to the queue elements at least to ease direct comparison to avoid repetition. Avoiding adding a child of an enqueued parent may be more involved. In the end, we'll have to see if it's worth the effort to add some id to the queue or using compareDocumentPosition(), contains() or similar.

Yeah, I thought something like this like using `contains()`, but it can be too expensive to do at the resource cost level and not the implementation. The position idea would also be a good idea, as well as the depth level help but the bad thing is that we don't have control of the application, so we would have to map everything before doing the analysis, and probably in each mutation we would have to update since new elements can appear or change the entire structure of the tree, so that would also be very expensive. But I think we can try to explore that at some point.

> There's also the option of removing from the queue any deleted node reported by a MutationRecord (including children). Again, I'm not sure if this would be better than just dequeue and check if node is in the doc before triggering the analysis...

Hmm, we keep the deleted nodes still in the queue, because when it runs, we check to see if it still exists before analysis, as it's a quick check, the scheduler can quickly jump to the next callback in the queue. So I think this is safer and faster in my point of view, it can have a large number of elements that do not exist in the queue but when they are executed this should not harm any element that is in the queue and has to be executed as soon as possible. The downside is that we can keep the element's reference in memory until it runs but it should be minimal, I imagine.

**Comment by matuzalemsteles** [comment]
I sent in a few more commits that add some axe-core configs to the settings and improvements to the mutation criteria, but I'm still investigating the issues with denylist in exclude, tomorrow I'll continue investigating this issue.

File: modules/apps/frontend-js/frontend-js-a11y-web/build.gradle [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+	compileOnly group: "org.apache.felix", name: "org.apache.felix.http.servlet-api", version: "1.1.2"
+	compileOnly group: "org.osgi", name: "org.osgi.service.component.annotations", version: "1.3.0"
+	compileOnly group: "org.osgi", name: "osgi.core", version: "6.0.0"
+	compileOnly project(":apps:frontend-js:frontend-js-loader-modules-extender-api")
+	compileOnly project(":apps:static:portal-configuration:portal-configuration-metatype-api")
+}
+
+jar {
+	enabled System.getenv('NODE_ENV') == "development"
+}
```
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: Just an observation about this implementation, this won't respect `nodejs.node.env` when we're tinkering with this module to deploy, so if you come to work on this module you need to set it to `NODE_ENV=development` in the console before calling the deploy.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		name = "axe-core-run-only-name", required = false
+	)
+	public String[] axeCoreRunOnly();
+
+	@Meta.AD(
+		deflt = "violations|inapplicable|passes|incomplete",
+		description = "axe-core-result-types-description",
+		name = "axe-core-result-types-name", required = false
+	)
+	public String[] axeCoreResultTypes();
```
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:67]: I have added this setting but I think the most ideal setting for this setting in the UI should be a multi select or checkbox list, not sure which type to use here for that.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java [hunk 6]
```diff
@@ -12,4 +12,22 @@
...
+/**
+ * @author Matuzalem Teles
+ */
+@ExtendedObjectClassDefinition(generateUI = false)
+@Meta.OCD(
+	id = "com.liferay.frontend.js.a11y.web.internal.configuration.FlagA11yConfiguration"
+)
+public @interface FlagA11yConfiguration {
+
+	@Meta.AD(deflt = "true", required = false)
```
Comment (matuzalemsteles) [inline modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java:30]: @dsanz I'm thinking about sending this disabled for the time being, because we still need the Panel that didn't come in, and also maybe we have to work on an adoption strategy? I'm just worried that one day developers will wake up and see products with too many accessibility errors, some products have more, some less, some are our own components that teams must configure or we can support.

**Review by matuzalemsteles — COMMENTED** [review]
I think this is ready for a final review, would love some java expert to take a look mainly to see the Liferay patterns 😅

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java (hunk 5 above)
Comment (dsanz) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:67]: I may have missed it, however, I glanced over tool code, and I _think_ we're reporting just the violations. So, I don't see how asking axe to give other results would be useful here. We may then omit this config item until we do something with the other result types. That may happen in future iterations if we prioritize report completeness

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java (hunk 6 above)
Comment (dsanz) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java:30]: I agree. I would officially release the a11y checker once side panel is ready, not before.

In addition, I think we should give ourselves (FI team) some time to test this in the portal. Reasons are:
- As this tool is targeted to developers, we as FI folks should get familiar with how tool works, what can we do with the info it gives, etc. This is a way to check how useful is what we created and get early feedback from target users.
- We'd like to see what's the perceived impact in performance when navigating in presence of the tool, particularly, in complex pages. This may lead to tweak the configurations or even make some deeper optimizations in the tool.
- We may find that some of the rule violations are caused by our own components. It'd be nice if we make sure no clay component or theme we own brings these issues, before tool is enabled by default. This would save valuable time and frustration to other product team devs.

I believe this will pay off, however, as these things require some effort (i.e. time) from our people, I'd like to agree on this plan with @nhpatt and @david-truong What do you guys think?

File: modules/apps/frontend-js/frontend-js-a11y-web/build.gradle (hunk 4 above)
Comment (dsanz) [reply modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: In addition to the bare env var, maybe we can check if the property was set via `build.<username>.properties `as well. I'm not sure this is a common pattern but it makes sense to me.

Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: Hmm, I think it's a good idea, but I have no idea how to implement this 😅, well, I can do some research and see what I can do

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java (hunk 5 above)
Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:67]: Hmm, I think it's a good idea, but I have no idea how to implement this 😅, well, I can do some research and see what I can do

Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/A11yConfiguration.java:67]: hmm yeah, I think I'm going to remove that from the config and just keep violations.

File: modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java (hunk 6 above)
Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/src/main/java/com/liferay/frontend/js/a11y/web/internal/configuration/FlagA11yConfiguration.java:30]: I totally agree, I think it makes a lot of sense to experiment first before releasing it to everyone and fix the issues with our components, both on the Clay and DXP sides.

So for now I will keep this disabled before I go.

File: modules/apps/frontend-js/frontend-js-a11y-web/build.gradle (hunk 4 above)
Comment (nhpatt) [reply modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: I haven't done it either but as a last resort you can read the file directly. I would guess that, being a properties file, it should be available somewhere like in project.properties: https://github.com/brianchandotcom/liferay-portal/blob/ea7567225b6382054d1128a0d0c82146b1acaa56/modules/releng.gradle#L116

Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: Hmm thank you @nhpatt, this will help

Comment (matuzalemsteles) [reply modules/apps/frontend-js/frontend-js-a11y-web/build.gradle:13]: I did some testing getting the `project.properties` property and it seems to work fine.

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
Hey @john-co or @nkkwon can you help me understand what happened with these tests, if it is something related to this PR?

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
ci:reevaluate:1257401_1171

**Comment by nkkwon** [comment]
Just started reviewing :)

**Comment by nkkwon** [comment]
**Analysis Details**

```Warning: Could not find file /opt/dev/projects/github/liferay-portal/bundles/osgi/modules/com.liferay.frontend.js.a11y.web.jar to copy.```  
[lpkg-base Jenkins Log](https://testray.liferay.com/reports/production/logs/2021-06/test-1-25/test-portal-acceptance-pullrequest(master)/1171/lpkg-base-jdk8/0/0/jenkins-console.txt.gz)
Looks like this is the main error for the PR.

**Comment by matuzalemsteles** [comment]
Oh thanks @nkkwon!

Hey @nhpatt and @dsanz, I think this is related to our rule in `jar`, maybe we can add this in some step before generating the `jar`? maybe in the build if possible.

**Comment by nhpatt** [comment]
Nice, there is a check in the build script...

I don't know... I feel that not generating the jar is the right choice. We could change the gradle code to generate an empty jar though. But, as this is supposed to be an internal tool should be marked with an `.lfrbuild-portal` ?

**Comment by dsanz** [comment]
> Nice, there is a check in the build script...
> 
> I don't know... I feel that not generating the jar is the right choice. We could change the gradle code to generate an empty jar though or change the script. But, as this is supposed to be an internal tool should be marked with an `.lfrbuild-portal` ? @dsanz

I think so, otherwise, it won't be deployed according to [this](https://github.com/liferay/liferay-portal/blob/master/modules/README.markdown#build)

> That marker file marks the module to be "bundled" with the portal and built by the CI, there are other marker files that allow the CI/release script to ignore the module but I don't know if they support the combination we want: to be deployed with `ant all` but ignored by CI/release script.

According to [this](https://github.com/liferay/liferay-portal/blob/master/modules/README.markdown#continuous-integration), th e combination we're looking for does not exist.

We still may want to add `.lfrbuild-releng-ignore` marker to prevent module to be included in the release, which is something we want. I don't think this will solve the CI issue, but we may try ...

If the above does not work, then we'll have to work at the build script level to either generate a dummy jar or packaging it in a way that it's disabled to the OSGi container eyes (I haven't seen a case where this needs to be done, we may be the first ones)

**Comment by matuzalemsteles** [comment]
Thanks @dsanz and @nhpatt, So, I'll add `.lfrbuild-releng-ignore`, just to test it now, but if it doesn't work, I'll probably investigate the other method. Maybe I need a java expert or someone with knowledge of OSGI 🙂

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by nhpatt** [comment]
Looks fine :D

If it does not work, although seems that we've cheated the CI, there are several options we can try:

1. With sourceSets we can change the source directory to remove all java files and serve an empty module

```
sourceSets {
	main {
		java { exclude "**/*" ...
```

2. With processResources we can remove resource files and try to remove the manifest:

```
processResources {
	exclude '**/*'
}
```

3. Or we can try configuring the manifest task inside jar { to disable or generate a dumb one
4. Or play with the bnd plugin that is generating the manifest to generate an empty manifest (https://github.com/bndtools/bnd/tree/master/biz.aQute.bnd.gradle)
5. Or play with the bundle tasks to try to disable it.
6. Or deploy a config file that adds the bundle to the blacklist directly (this is easy and should work)

Either way this 6 options work by generating a jar and then disabling it... something that we can do with a configuration directly, right? I would prefer to *not* generate the jar if possible (if everything works as apparently works right now).

**Comment by matuzalemsteles** [comment]
@nhpatt yes, surprised that passed as well 😅 well, I think I also prefer not to create a jar to simulate disabling the module. I think if we manage not to push this out for release and we manage to enable it in the development environment, that looks great, do you guys see any issues with continuing with this or do you think we should test more before submitting? This is currently also being disabled by default, even though it is in the development environment, we can probably enable it when we submit this https://issues.liferay.com/browse/LPS-134118 or it will depend on what we decide on the plan to enable this.

**Comment by dsanz** [comment]
> @nhpatt yes, surprised that passed as well  well, I think I also prefer not to create a jar to simulate disabling the module. I think if we manage not to push this out for release and we manage to enable it in the development environment, that looks great, do you guys see any issues with continuing with this or do you think we should test more before submitting? This is currently also being disabled by default, even though it is in the development environment, we can probably enable it when we submit this https://issues.liferay.com/browse/LPS-134118 or it will depend on what we decide on the plan to enable this.

My opinion on this is [somewhere above this thread](https://github.com/liferay-frontend/liferay-portal/pull/1146#discussion_r648883258). I believe we should open an internal adoption period where we can test the tool internally. During that time we can finish the side panel in parallel. I'm afraid tool will be too obstrusive and that may affect adoption, hence this period of fine-tuning and evaluation. So, basically, tool would need to be enabled manually: besides the node `env` variable, we'd need to add the osgi `.config` file to turn the feature on.

**Comment by matuzalemsteles** [comment]
> My opinion on this is somewhere above this thread. I believe we should open an internal adoption period where we can test the tool internally. During that time we can finish the side panel in parallel. I'm afraid tool will be too obstrusive and that may affect adoption, hence this period of fine-tuning and evaluation. So, basically, tool would need to be enabled manually: besides the node env variable, we'd need to add the osgi .config file to turn the feature on.

yeah, I agree on that too, we currently need to add a `.config` to enable it.

**Comment by matuzalemsteles** [comment]
Well, I'll wait for the PR of @diegonvs from the Panel to enter the master so I can resolve conflicts before submitting it.

**Comment by matuzalemsteles** [comment]
ci:test:sf

**Comment by matuzalemsteles** [comment]
ci:test:relevant

**Comment by matuzalemsteles** [comment]
Hey @nhpatt and @dsanz I think its ready to go, as only you guys have commented, let me know if there's anything else we should consider before going ahead with this.

**Comment by nhpatt** [comment]
LGTM :)

**Comment by matuzalemsteles** [comment]
Ok, So, I go ahead! @dsanz if you have something in mind, let me know and I'll work on another RP if necessary

**Comment by matuzalemsteles** [comment]
ci:forward
